### PR TITLE
Refactor cpd

### LINKED ISSUE #595: Refactor TabularCPD
Currently `TabularCPD` has the these attributes: `variable`, `variable_card`, `evidence`, `evidence_card`.
And since it inherits `Factor`, it also has: `variables`, `cardinality` and `values`. 

`evidence` and `evidence_card` can be easily derived if we know other attributes. And since we are not directly using `evidence` or `evidence_card` too many times, we can remove these two and compute these when needed. Right now `evidence` and `evidence_card` is adding overhead in a lot of operations because we need to update these as well as `variables` and `cardinality`.

### PR DESCRIPTION
**Pull request by anaviltripathi**
I have removed `evidence` and `evidence_card` as was discussed in #595. I have also modified the tests accordingly. Let me know if there are some issues. 
Fixes #595 and #627.

### CODE REVIEW

File: pgmpy/factors/CPD.py [hunk 1]
```diff
@@ -11,6 +12,7 @@
 from pgmpy.extern import tabulate
 from pgmpy.extern import six
 from pgmpy.extern.six.moves import range, zip
+from warnings import warn
```
Comment (ankurankan) [inline pgmpy/factors/CPD.py:15]: @anaviltripathi Check PEP 8 for grouping import statements: https://www.python.org/dev/peps/pep-0008/#imports

File: pgmpy/factors/CPD.py [hunk 2]
```diff
@@ -107,27 +107,25 @@ def __init__(self, variable, variable_card, values,
...
         if evidence is not None:
-            if not isinstance(evidence, (list, set, tuple)):
+            if not isinstance(evidence, (list, tuple)):
                 if isinstance(evidence, np.ndarray):
                     evidence = evidence.tolist()
                 elif isinstance(evidence, six.string_types):
                     evidence = [evidence]
                 else:
-                    raise TypeError("Evidence must be list, set, tuple or array"
+                    raise TypeError("Evidence must be list, tuple or array"
```
Comment (ankurankan) [inline pgmpy/factors/CPD.py:127]: @anaviltripathi Must be a list, tuple or array of variable names and variable name can be any hashable python object.

File: pgmpy/factors/CPD.py [hunk 3]
```diff
@@ -107,27 +107,25 @@ def __init__(self, variable, variable_card, values,
...
                 if isinstance(evidence_card, np.ndarray):
                     evidence_card = evidence_card.tolist()
                 elif isinstance(evidence_card, (int, float)):
                     evidence_card = [evidence_card]
                 else:
-                    raise TypeError("Evidence cardinality must be list, set, "
+                    raise TypeError("Evidence cardinality must be list, "
                                     "tuple, numpy ndarray or simply a number.")
-            self.evidence_card = evidence_card
             cardinality.extend(evidence_card[::-1])
```
Comment (ankurankan) [inline pgmpy/factors/CPD.py:118]: @anaviltripathi Why are you reversing the lists ? Wouldn't it assign wrong values to different states ?

Comment (anaviltripathi) [inline pgmpy/factors/CPD.py:118]: @ankurankan Actually, this is part of the old code. Previous implementation used reversing the lists so I didn't modify it. I believe it has to do with [reshaping](https://github.com/pgmpy/pgmpy/blob/dev/pgmpy/factors/Factor.py#L113) the array according to cardinalities, and I think it is correct but counter-intuitive at first glance. Let me know if I am wrong.

File: pgmpy/factors/CPD.py (hunk 2 above)
Comment (anaviltripathi) [inline pgmpy/factors/CPD.py:127]: @ankurankan Ok, I will add this sentence.

File: pgmpy/factors/CPD.py (hunk 1 above)
Comment (anaviltripathi) [inline pgmpy/factors/CPD.py:15]: @ankurankan Yes, about this, actually there is one more issue you might have noticed. Due to carelessness I created a branch that had previous commits with regards to other PRs I had sent and failed to create a fresh branch. Although I have only kept the features pertaining to this particular feature, I missed this line. If you want, I could start afresh on a new branch that doesn't have irrelevant commit history.

File: pgmpy/factors/CPD.py (hunk 3 above)
Comment (ankurankan) [inline pgmpy/factors/CPD.py:118]: @anaviltripathi Actually we need to just make this point clear in our docs. For example with the implementation that you have done:

``` python
 cpd = TabularCPD('A', 2, [[0.1, 0.2, 0.3, 0.4], [0.9, 0.8, 0.7, 0.6]], ['B', 'C'], [2, 2])
```

this results in a CPD like this:

```
╒═════╤═════╤═════╤═════╤═════╕
│ C   │ C_0 │ C_0 │ C_1 │ C_1 │
├─────┼─────┼─────┼─────┼─────┤
│ B   │ B_0 │ B_1 │ B_0 │ B_1 │
├─────┼─────┼─────┼─────┼─────┤
│ A_0 │ 0.1 │ 0.2 │ 0.3 │ 0.4 │
├─────┼─────┼─────┼─────┼─────┤
│ A_1 │ 0.9 │ 0.8 │ 0.7 │ 0.6 │
╘═════╧═════╧═════╧═════╧═════╛
```

And without the reversing it would have been:

```
╒═════╤═════╤═════╤═════╤═════╕
│ B   │ B_0 │ B_0 │ B_1 │ B_1 │
├─────┼─────┼─────┼─────┼─────┤
│  C   │ C_0 │ C_1 │ C_0 │ C_1 │
├─────┼─────┼─────┼─────┼─────┤
│ A_0 │ 0.1 │ 0.2 │ 0.3 │ 0.4 │
├─────┼─────┼─────┼─────┼─────┤
│ A_1 │ 0.9 │ 0.8 │ 0.7 │ 0.6 │
╘═════╧═════╧═════╧═════╧═════╛
```

So, we need to make this clear in the docs that how does the order of evidence affect the parents of the CPD.

And the previous behavior was the latter one. I think we should keep that only as that's more intuitive. What do you think ?

File: pgmpy/factors/CPD.py (hunk 1 above)
Comment (ankurankan) [inline pgmpy/factors/CPD.py:15]: @anaviltripathi Please start on a fresh branch. And you can [`cherry-pick`](https://git-scm.com/docs/git-cherry-pick) the relevant commits from this branch.

File: pgmpy/factors/CPD.py (hunk 3 above)
Comment (anaviltripathi) [inline pgmpy/factors/CPD.py:118]: @ankurankan I am a little confused here. As I have already mentioned I didn't change these lines and removing `evidence` and `evidence_card` shouldn't have affected this as values being entered in the CPD are actually entered by calling its super class `Factor`. 
Also, although when using `print(cpd)` to print the cpd in my case gives weird output that I don't really understand why, below are some other things that I will need clarification upon.

Previous version, which is currently in use:

``` python
>>> cpd = TabularCPD('A', 2, [[0.1, 0.2, 0.3, 0.4, 0.01, 0.02], [0.9, 0.8, 0.7, 0.6, 0.99, 0.98]], ['B', 'C'], [3, 2])
>>> cpd.values
array([[[ 0.1 ,  0.2 ,  0.3 ],
        [ 0.4 ,  0.01,  0.02]],

       [[ 0.9 ,  0.8 ,  0.7 ],
        [ 0.6 ,  0.99,  0.98]]])
>>> 
KeyboardInterrupt
>>> print(cpd)
+-----+-----+-----+-----+-----+-----+-----+
| B   | B_0 | B_0 | B_1 | B_1 | B_2 | B_2 |
+-----+-----+-----+-----+-----+-----+-----+
| C   | C_0 | C_1 | C_0 | C_1 | C_0 | C_1 |
+-----+-----+-----+-----+-----+-----+-----+
| A_0 | 0.1 | 0.2 | 0.3 | 0.4 | 0.0 | 0.0 |
+-----+-----+-----+-----+-----+-----+-----+
| A_1 | 0.9 | 0.8 | 0.7 | 0.6 | 0.9 | 0.9 |
+-----+-----+-----+-----+-----+-----+-----+
>>> cpd.variables
['A', 'C', 'B']
>>> cpd.cardinality
array([2, 2, 3])
```

My current version which reverses the list (as I believe was done previously too):

``` python
>>> cpd = TabularCPD('A', 2, [[0.1, 0.2, 0.3, 0.4, 0.01, 0.02], [0.9, 0.8, 0.7, 0.6, 0.99, 0.98]], ['B', 'C'], [3, 2])
>>> cpd.values
array([[[ 0.1 ,  0.2 ,  0.3 ],
        [ 0.4 ,  0.01,  0.02]],

       [[ 0.9 ,  0.8 ,  0.7 ],
        [ 0.6 ,  0.99,  0.98]]])
>>> print(cpd)
+-----+-----+-----+-----+-----+-----+-----+
| C   | C_0 | C_0 | C_0 | C_1 | C_1 | C_1 |
+-----+-----+-----+-----+-----+-----+-----+
| B   | B_0 | B_1 | B_2 | B_0 | B_1 | B_2 |
+-----+-----+-----+-----+-----+-----+-----+
| A_0 | 0.1 | 0.2 | 0.3 | 0.4 | 0.0 | 0.0 |
+-----+-----+-----+-----+-----+-----+-----+
| A_1 | 0.9 | 0.8 | 0.7 | 0.6 | 0.9 | 0.9 |
+-----+-----+-----+-----+-----+-----+-----+
>>> cpd.variables
['A', 'C', 'B']
>>> cpd.cardinality
array([2, 2, 3])

```

If don't reverse the list, as suggested by you:

``` python
>>> cpd = TabularCPD('A', 2, [[0.1, 0.2, 0.3, 0.4, 0.01, 0.02], [0.9, 0.8, 0.7, 0.6, 0.99, 0.98]], ['B', 'C'], [3, 2])
>>> cpd.values
array([[[ 0.1 ,  0.2 ],
        [ 0.3 ,  0.4 ],
        [ 0.01,  0.02]],

       [[ 0.9 ,  0.8 ],
        [ 0.7 ,  0.6 ],
        [ 0.99,  0.98]]])
>>> print(cpd)
+-----+-----+-----+-----+-----+-----+-----+
| B   | B_0 | B_0 | B_1 | B_1 | B_2 | B_2 |
+-----+-----+-----+-----+-----+-----+-----+
| C   | C_0 | C_1 | C_0 | C_1 | C_0 | C_1 |
+-----+-----+-----+-----+-----+-----+-----+
| A_0 | 0.1 | 0.2 | 0.3 | 0.4 | 0.0 | 0.0 |
+-----+-----+-----+-----+-----+-----+-----+
| A_1 | 0.9 | 0.8 | 0.7 | 0.6 | 0.9 | 0.9 |
+-----+-----+-----+-----+-----+-----+-----+
>>> cpd.variables
['A', 'B', 'C']
>>> cpd.cardinality
array([2, 3, 2])

```

Notice that in my case `cpd.values`, `cpd.variables`, `cpd.cardinality` match with the previous implementation but not the output with `print(cpd)`. 
So, I am reversing the lists as I think that values should be in the same format as they were previously for expected behavior of TabularCPD. 
I maybe completely wrong here. Let me know.

Comment (ankurankan) [inline pgmpy/factors/CPD.py:118]: @anaviltripathi What I understand now (not very sure) is that the earlier representation was considering the fastest changing parent first. So, if the evidence was `['B', 'C']` then the table should have been something like this:

```
+-----+-----+-----+-----+-----+-----+-----+
| C   | C_0 | C_1 | C_0 | C_1 | C_0 | C_1 |
+-----+-----+-----+-----+-----+-----+-----+
| B   | B_0 | B_0 | B_1 | B_1 | B_2 | B_2 |
+-----+-----+-----+-----+-----+-----+-----+
+-----+-----+-----+-----+-----+-----+-----+
```

But the `__str__` method wasn't considering this.

Anyways let's fix this all now. I think the earlier representation of evidence was not that intuitive. And intuitive way for defining evidence for the above table should be `['C', 'B']`. What do you think ?

Comment (ankurankan) [inline pgmpy/factors/CPD.py:118]: @anaviltripathi What I understand now (not very sure) is that the earlier representation was considering the fastest changing parent first. So, if the evidence was `['B', 'C']` then the table should have been something like this:

```
+-----+-----+-----+-----+-----+-----+-----+
| C   | C_0 | C_1 | C_0 | C_1 | C_0 | C_1 |
+-----+-----+-----+-----+-----+-----+-----+
| B   | B_0 | B_0 | B_1 | B_1 | B_2 | B_2 |
+-----+-----+-----+-----+-----+-----+-----+
+-----+-----+-----+-----+-----+-----+-----+
```

But the `__str__` method wasn't considering this.

Anyways let's fix this all now. I think the earlier representation of evidence was not that intuitive. And intuitive way for defining evidence for the above table should be `['C', 'B']`. What do you think ?

Comment (anaviltripathi) [inline pgmpy/factors/CPD.py:118]: @ankurankan I agree with you. So, I am not reversing the list finally and the output would be as in the third case. Also, in many other places, just like in `__str__` method, `evidence` was used directly, so the method implementation and their tests were designed that way only. I may need to look into those carefully and I will send in patches as I keep working on those to make sure new bugs are not introduced. 
In short, new implementation will be they way you suggested above.

Comment (anaviltripathi) [inline pgmpy/factors/CPD.py:118]: I have one more doubt. Should we allow reducing or marginalizing the variable itself in CPD, especially inplace. If it were a JPD then marginalizing wrt any variable would be correct, but I suspect that it is not the case here. 
Currently reducing and marginalizing w.r.t. `variable` is allowed and it gives all `1.` output as can be expected if this behavior is to be allowed. But, `print(cpd)` doesn't work as can also be expected. 
But, in my view this should not be allowed inplace. If inplace is chosen to `False` then it can be allowed I guess. 
Also, docstrings for marginalize and reduce are not correct, errors pop up if you follow the directions exactly.

**Comment by yashu-seth** [comment]
@anaviltripathi You mean error pops up in this branch or in the dev for marginalize and reduce?

**Comment by anaviltripathi** [comment]
@yashu-seth In this branch, and in dev.

File: pgmpy/factors/CPD.py (hunk 3 above)
Comment (anaviltripathi) [inline pgmpy/factors/CPD.py:118]: Update: I ran tests by not reversing list of evidence and a lot of tests are failing (30). Should we really opt for not reversing the lists, considering that it has repercussions in many other methods?

**Comment by yashu-seth** [comment]
@anaviltripathi Okay. I think you have opened issue #635 for it.

**Comment by anaviltripathi** [comment]
@yashu-seth No, that is different. What I am talking about here is whether we should even allow reducing and marginalizing with respect to the variable under consideration in a TabularCPD. Present implementation allows this, but I think that we should not allow it inplace, and maybe give a warning even for case of non-inplace. 
Also, even though current implementation allows this, after reducing w.r.t. the variable `print` command doesn't work and produces errors. I didn't think necessary to open a separate issue for this. Should I?

**Comment by yashu-seth** [comment]
@anaviltripathi I also got errors for the cases you mentioned. I think issue #635 is similar to this problem because I get similar error for both. We will have to look into it to be sure. I think if not a separate issue we must mention our findings there itself.

**Comment by anaviltripathi** [comment]
This branch had some irrelevant commit history so I am closing this. Refer new PR #643.
